rect: enhanced rounding and chamfer options

hvegh opened the pull request:
Introduces the extra "cswap" parameter that moves the chamfer and rounding center points.

In the original code the negative roudings and chamfers are only extended in the x direction, this patch allows control over both directions.

Example:

$fn = 32;
translate([0,0,-.51])
distribute(spacing=1.5) {

// rounded examples:

    // original
    rect(rounding=.25*[1,1,-1,-1]);

    // cswap curve for quadrants [-1,1] and [-1,-1]
    rect(rounding=.25*[1,1,-1,-1], cswap = [0,1,1,0]);

// chamfered examples:

    // original
    rect(chamfer=.25*[1,1,-1,-1]);

    // cswap chamfer for quadrants [-1,1] and [-1,-1]
    // note only affects the negative chamfers
    rect(chamfer=.25*[1,1,-1,-1], cswap = [0,1,1,0]);
}
![rect_enhance](https://github.com/BelfrySCAD/BOSL2/assets/4540127/e1c9ca22-d068-4b9d-8600-0fc571f44d9e)

adrianVmariano commented:
I don't like the cswap parameter name.  Not sure what it should be.  corner_flip?

hvegh commented:
> I don't like the cswap parameter name. Not sure what it should be. corner_flip?

I am horrible at thinking of sensible names, corner_flip sounds perfect. 
If that is the consensus, I am happy to change it I am happy.

Also note that a corner flip for a specific quadrant is implemented as a number: zero -> no_flip  /  non zero -> flip.

I was wondering if this should be boolean type, any preference?

adrianVmariano commented:
Coming up with names can be pretty difficult.  We sometimes struggle.  Go ahead with corner_flip.  

The corner_flip parameter should definitely be boolean, and it should be permitted to give a list or give it as a single boolean, so true expands into [true,true,true,true].  

Also did you confirm that the anchor overrides work properly for the case when atype=="perim"?

hvegh commented:
Update:

- fixed anchor overrides when atype = "perim" 
- renamed to corner_flip
- added some comments

```
flip    = true;
chamfer = 5*[0,-1,1,0];
rounding= 5*[-1,0,0,1];

distribute(spacing=40) {
    rect(size=20, atype="perim", chamfer = chamfer, rounding=rounding, corner_flip = flip)
        show_anchors();

    rect(size=20, atype="perim", chamfer = -chamfer, rounding=-rounding, corner_flip = !flip)
        show_anchors();

    rect(size=20, atype="perim", chamfer = -chamfer, rounding=-rounding, corner_flip = flip)
        show_anchors();
}
```
![rect_test_perim](https://github.com/BelfrySCAD/BOSL2/assets/4540127/b6857b01-1092-40a7-abb7-d02797545674)
